test/internal/layers/lazy.go

```diff
@@ -0,0 +1,166 @@
+//go:build windows
+
+package layers
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/google/go-containerregistry/pkg/crane"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/Microsoft/hcsshim/ext4/tar2ext4"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/security"
+	"github.com/Microsoft/hcsshim/internal/wclayer"
+	"github.com/Microsoft/hcsshim/pkg/ociwclayer"
+
+	"github.com/Microsoft/hcsshim/test/internal/constants"
+	"github.com/Microsoft/hcsshim/test/internal/util"
+)
+
+// helper utilities for dealing with images
+
+type LazyImageLayers struct {
+	Image    string
+	Platform string
+	TempPath string // TempPath is the path to create a temporary directory in. Default in [os.TempDir]
+	once     sync.Once
+	layers   []string
+}
+
+// ImageLayers returns the image layer paths, from lowest to highest, for a particular image.
+func (x *LazyImageLayers) ImageLayers(ctx context.Context, tb testing.TB) []string {
+	tb.Helper()
+	tb.Logf("pulling and unpacking %s image %q", x.Platform, x.Image)
+	// don't use tb.Error/Log inside Once.Do stack, since we cannot call tb.Helper before executing f()
+	// within Once.Do and that will therefore show the wrong stack/location
+	var err error
+	x.once.Do(func() {
+		var dir string
+		// tb.TempDir is deleted at the end of a test, but we want the image for future test runs
+		dir, err = os.MkdirTemp(x.TempPath, util.CleanName(x.Image))
+		if err != nil {
+			err = fmt.Errorf("failed to create temp directory: %w", err)
+			return
+		}
+
+		switch x.Platform {
+		case constants.PlatformLinux:
+			err = x.linuxImage(ctx, dir)
+		case constants.PlatformWindows:
+			err = x.windowsImage(ctx, dir)
+		default:
+			err = fmt.Errorf("unsupported platform %q", x.Platform)
+		}
+	})
+	if err != nil {
+		x.Close(ctx)
+		tb.Fatal(err)
+	}
+	return x.layers
+}
+
+// Close removes the downloaded image layers.
+func (x *LazyImageLayers) Close(ctx context.Context) {
+	for _, dir := range x.layers {
+		d, err := filepath.Abs(dir)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("count not get absolute path to %q", dir)
+			continue
+		}
+
+		if _, err := os.Stat(d); err != nil {
+			if !os.IsNotExist(err) {
+				log.G(ctx).WithError(err).Errorf("path %q is not valid", d)
+			}
+			continue
+		}
+		if err := wclayer.DestroyLayer(ctx, d); err != nil {
+			log.G(ctx).WithError(err).Errorf("could not destroy layer %q", d)
+		}
+	}
+}
+
+func (x *LazyImageLayers) linuxImage(ctx context.Context, dir string) error {
```

Comment:
Do we not need to handle the case of multiple layers for LCOW images like how we do for WCOW below?

Reply:
so [`crane.Export`](https://pkg.go.dev/github.com/google/go-containerregistry@v0.11.0/pkg/crane#Export) actually merges all compressed layers into a single uncompressed tar stream.
I didnt want to use it with WCOW images, since I think it would handle windows paths & the hives properly.

Reply:
But does this mean we won't be able to have any functional tests for LCOW that rely on having multiple image layers? (since all of the layers would be merged into a single layer)

Reply:
ohhh, true
we would need to add custom code to deal with that separately.
Do you know if there are tests in functional that do?

Reply:
I am not sure if we have any such tests but I imagine we would want to add such tests sooner or later. So I think if it is easy to change this function to create multiple layers we should do it. What do you think?

Reply:
done, its actually cleaner this way too (though the commit changes a lot ...)